ocaml/xapi-guard/src/main.ml

```diff
@@ -215,55 +230,86 @@ let vtpm_get_contents _dbg vtpm_uuid =
   @@ let* self = Server_interface.with_xapi ~cache @@ VTPM.get_by_uuid ~uuid in
      Server_interface.with_xapi ~cache @@ VTPM.get_contents ~self
 
-let rpc_fn =
+let rpc_fn ~vtpm_write_push ~uefi_write_push =
   let module Server =
     Xapi_idl_guard_privileged.Interface.RPC_API (Rpc_lwt.GenServer ()) in
   (* bind APIs *)
-  Server.varstore_create depriv_varstored_create ;
+  Server.varstore_create (depriv_varstored_create uefi_write_push) ;
   Server.varstore_destroy depriv_varstored_destroy ;
-  Server.vtpm_create depriv_swtpm_create ;
+  Server.vtpm_create (depriv_swtpm_create vtpm_write_push) ;
   Server.vtpm_destroy depriv_swtpm_destroy ;
   Server.vtpm_set_contents vtpm_set_contents ;
   Server.vtpm_get_contents vtpm_get_contents ;
   Rpc_lwt.server Server.implementation
 
-let process body =
+let process ~vtpm_write_push ~uefi_write_push body =
   let+ response =
     Xapi_guard.Dorpc.wrap_rpc Xapi_idl_guard_privileged.Interface.E.error
       (fun () ->
         let call = Jsonrpc.call_of_string body in
         D.debug "Received request from message-switch, method %s" call.Rpc.name ;
-        rpc_fn call
+        rpc_fn ~vtpm_write_push ~uefi_write_push call
     )
   in
   Jsonrpc.string_of_response response
 
+let retry_forever fname f =
+  let rec loop () =
+    let* () =
+      Lwt.catch f (function exn ->
+          D.info "%s failed with %s, retrying..." fname (Printexc.to_string exn) ;
+          Lwt_unix.sleep 0.5
```

Comment:
These retry functions are used in different situation. The exponential backoff is useful for expected errors during runtime, in particular to wait while xapi is offline and the pushes are expected to fail.

This function is a recovery mechanism in case there's a malfunction in the code: code is not expected to fail and reach the recovery mechanism during normal operation. Maybe there's a case for exponential backoff to avoid spamming the logs, or maybe xapi-guard should be restarted to try to recover, it's not clear to know in foresight